### ISSUE TITLE
Private type exposed in public without a warning

### ISSUE THREAD

**Issue by kvark** [rust-lang/rust#73410]
Here is the code:

```rust
pub mod foo {
    mod sub {
        pub struct Bar;
    }
    
    use sub::Bar;
    
    pub struct Foo {
        pub bar: Bar,
    }
}

//use foo::Bar; // this errors for `Bar` being private
```

I expected to see this happen (compile error):
> Private type `sub::Bar` exposed in public signature of `Foo::bar`

Instead, this happened: compiles just fine.
Trying to use the type fails, since it's private

### Meta
Tested on stable 1.44.0 and nightly of 1.46.0

[playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=2c156df298f8ea8b992ee4a96220d7d5)

**Comment by cuviper** [comment]
Sometimes this is done intentionally, like the `Sealed` trait here:
https://rust-lang.github.io/api-guidelines/future-proofing.html

**Comment by Dylan-DPC** [comment]
Closing this as it is intended and not sure if there's anything that can be done here

**Comment by petrochenkov** [comment]
This is now addressed by a lint called `unnameable_types` (#48054), but that lint is allowed by default.
